### PR TITLE
implement inheritance

### PR DESCRIPTION
**Pull request by jyoo980**
**What is this?**
- [x] add a `NodeTable` to store all the nodes of Interfaces/Classes we create (of type AstNode)
- [x] add a new AST traversal phase for implementing inheritance and making the classes/interfaces fulfill their contracts
- [x] implement `fulfillContract` for `InterfaceDecl`

### CODE REVIEW

File: src/ast/ClassDecl.ts [hunk 1]
```diff
@@ -74,6 +75,23 @@ export class ClassDecl extends Content {
         this.functions.forEach((funcDecl: FuncDecl) => funcDecl.typeCheck());
     }
 
+    public fulfillContract(): void {
```
Comment (scveloso) [inline src/ast/ClassDecl.ts:106]: if class A extends abstract class B, then A needs to implement B's methods right? this case seems to be missing

**Review by scveloso — COMMENTED** [review]
LGTM overall, noticed one thing but tbh I'm not 100% sure myself

File: src/ast/ClassDecl.ts (hunk 1 above)
Comment (jyoo980) [reply src/ast/ClassDecl.ts:106]: Only if those methods are declared `abstract` - I'll check and see if we support abstract function declarations in our DSL.

Comment (jyoo980) [reply src/ast/ClassDecl.ts:106]: Nope!

We only support abstract classes, but (strangely) LOL - `FuncDecl` doesn't allow for `abstract`.

Comment (scveloso) [reply src/ast/ClassDecl.ts:106]: ok everythings gucci

**Review by scveloso — APPROVED** [review]

File: src/ast/ClassDecl.ts (hunk 1 above)
Comment (kiyomih) [reply src/ast/ClassDecl.ts:106]: nice to have? (abstract functions)

**Review by kiyomih — APPROVED** [review]
🔥🙌🏻💯

File: src/ast/ClassDecl.ts (hunk 1 above)
Comment (jyoo980) [reply src/ast/ClassDecl.ts:106]: @kiyomih sounds like it

**Comment by jyoo980** [comment]
@scveloso @kiyomih 

I'll merge this in after we have MVP on master

**Review by kiyomih — APPROVED** [review]
🔥🔥🔥

**Pull request merged**